source/bigint.c

```diff
@@ -448,254 +456,287 @@ int aws_bigint_bytebuf_debug_output(const struct aws_bigint *bigint, struct aws_
-        non_zero_high_digit_bytes = 3;
-    }
-
-    return (digit_count - 1) * BYTES_PER_BASE_DIGIT + non_zero_high_digit_bytes;
-}
-
-#if _MSC_VER
-#    pragma warning(push)
-#    pragma warning(disable : 4221) /* allow struct member to reference format_buffer  */
-#    pragma warning(disable : 4204) /* non-constant aggregate initializer */
-#endif
-
 int aws_bigint_bytebuf_append_as_big_endian(
     const struct aws_bigint *bigint,
     struct aws_byte_buf *buffer,
-    size_t minimum_length) {
+    size_t desired_bit_width) {
 
-    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
-    AWS_PRECONDITION(aws_byte_buf_is_valid(buffer));
+    AWS_FATAL_PRECONDITION((desired_bit_width % BITS_PER_BYTE) == 0);
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
+    AWS_FATAL_PRECONDITION(aws_byte_buf_is_valid(buffer));
 
     size_t digit_count = aws_array_list_length(&bigint->digits);
-    size_t bigint_bytes = s_aws_bigint_byte_length(bigint);
-    size_t padding_bytes = 0;
-    if (bigint_bytes < minimum_length) {
-        padding_bytes = minimum_length - bigint_bytes;
+    size_t byte_length = digit_count * BYTES_PER_BASE_DIGIT;
+    if (desired_bit_width != 0) {
+        byte_length = (desired_bit_width - 1) / BITS_PER_BYTE + 1;
     }
 
     int result = AWS_OP_ERR;
-
-    size_t required_capacity = bigint_bytes + padding_bytes;
-    if (aws_byte_buf_reserve_relative(buffer, required_capacity)) {
+    if (aws_byte_buf_reserve_relative(buffer, byte_length)) {
         goto done;
     }
 
-    if (padding_bytes > 0) {
-        uint8_t padding = 0;
-        struct aws_byte_cursor padding_cursor = {
-            .ptr = &padding,
-            .len = 1,
-        };
-
-        for (size_t i = 0; i < padding_bytes; ++i) {
-            if (aws_byte_buf_append(buffer, &padding_cursor)) {
-                goto done;
-            }
-        }
+    uint8_t *raw_buffer = buffer->buffer + buffer->len;
+    for (size_t i = 0; i < byte_length; ++i) {
+        *(raw_buffer + i) = 0;
     }
+    buffer->len += byte_length;
 
-    uint8_t digit_byte = 0;
-    struct aws_byte_cursor digit_bytes_cursor = {
-        .ptr = &digit_byte,
-        .len = 1,
-    };
-
-    bool should_write = false;
-
-    for (size_t i = 0; i < digit_count; ++i) {
-        size_t digit_index = digit_count - i - 1;
-
-        uint32_t current_digit = 0;
-        aws_array_list_get_at(&bigint->digits, &current_digit, digit_index);
-
-        for (size_t j = 0; j < BYTES_PER_BASE_DIGIT; ++j) {
-            digit_byte = (uint8_t)(current_digit >> (3 * BITS_PER_BYTE)) & ((1U << BITS_PER_BYTE) - 1);
-            current_digit <<= BITS_PER_BYTE;
+    size_t bytes_to_write = byte_length;
+    if (bytes_to_write > digit_count * BYTES_PER_BASE_DIGIT) {
+        bytes_to_write = digit_count * BYTES_PER_BASE_DIGIT;
+    }
 
-            if (!should_write) {
-                should_write = digit_byte > 0 || j + 1 == BYTES_PER_BASE_DIGIT;
-            }
+    /* dumb/wasteful but simple */
```

Comment:
I bet this actually optimizes really well

Reply:
Actually, come to think of it, if this vectorizes, it's going to not be constant time with the C impl. Does that matter? Do we need to disable or force vectorization here?

Reply:
This one needs to be constant-time.  Maybe throwing volatile at it will help.